File: structures/stack/javascript/__tests__/stack.test.js
import { Stack } from "../stack/stack.js";

import assert from "assert";

/*
Data for testing all stack functions
 */
const TEST_DATA1 = [1, 2, 3, 4, 5, 6, 7];
const TEST_DATA2 = ["a", "e", "b", "d", "f"];
const TEST_DATA3 = [7.5, 5.3, 1.2, 4.7, 15.2, 6.3, 7.4];

/*
Data for testing stack function pop
 */
const TEST_DATA_POP1 = [1, 2, 3, 4]; // -3
const TEST_DATA_POP2 = ["a", "e", "b"]; // -2
const TEST_DATA_POP3 = [7.5, 5.3, 1.2, 4.7, 15.2, 6.3]; //- 1


/*
Helped function for testing the push stack function 
 */
function testPush(array){
    let stack = new Stack();

    array.forEach(element => {
        stack.push(element);
    });
    return stack.array_
}


/*
Helped function for testing the pop stack function 
 */
function testPop(array, size_pop){
    let stack = new Stack();

    array.forEach(element => {
        stack.push(element);
    });

    for (let i = 0; i < size_pop; i++)
        stack.pop();
    return stack.array_;
}

/*
Helped function for testing the top stack function 
 */
function testTop(array){
    let stack = new Stack();
    array.forEach(element => {
        stack.push(element);
    });
    let temp = stack.top();
    return temp;
}

/*
Helped function for testing the size stack function 
 */
function testSize(array){
    let stack = new Stack();

    array.forEach(element => {
        stack.push(element);
    });
    return stack.size();
}

/*
Helped function for testing the empty stack function 
 */
function testEmpty(array){
    let stack = new Stack();

    array.forEach(element => {
        stack.push(element);
    });
    return stack.empty();
    
}

/*
Helped function for testing the swap stack function 
 */
function testSwap(array1, array2) {
    var stack1 = new Stack();
    array1.forEach(element => {
        stack1.push(element);
    });

    var stack2 = new Stack();
    array2.forEach(element => {
        stack2.push(element);
    });
    
    stack1.swap(stack2);
    
    return stack1.array_;
}


/*
Testing the swap stack function
*/
it('Test swap with TEST_DATA1 and TEST_DATA3', () => {
    var a = TEST_DATA1;
    var b = TEST_DATA3;
    let result = testSwap(a, b);
    assert.deepEqual(result, TEST_DATA3);
})

it('Test swap with TEST_DATA3 and TEST_DATA1', () => {
    var a = TEST_DATA3;
    var b = TEST_DATA1;
    let result = testSwap(a, b);
    assert.deepEqual(result, TEST_DATA1);
})


/*
Testing the empty stack function
*/
it('Test empty function with TEST_DATA1', () => {
    assert.equal(testEmpty(TEST_DATA1), false);
})

it('Test empty function with TEST_DATA2', () => {
    assert.equal(testEmpty(TEST_DATA2), false);
})

it('Test empty function with TEST_DATA3', () => {
    assert.equal(testEmpty(TEST_DATA3), false);
})

it('Test empty function with []', () => {
    assert.equal(testEmpty([]), true);
})


/*
Testing the size stack function
*/
it('Test size function with TEST_DATA1', () => {
    assert.equal(testSize(TEST_DATA1), TEST_DATA1.length);
})
it('Test size function with TEST_DATA2', () => {
    assert.equal(testSize(TEST_DATA2), TEST_DATA2.length);
})
it('Test size function with TEST_DATA3', () => {
    assert.equal(testSize(TEST_DATA3), TEST_DATA3.length);
})


/*
Testing the push stack function
*/
it('Test push function with TEST_DATA1', () => {
    assert.deepEqual(testPush(TEST_DATA1), TEST_DATA1);
})
it('Test push function with TEST_DATA2', () => {
    assert.deepEqual(testPush(TEST_DATA2), TEST_DATA2);
})
it('Test push function with TEST_DATA3', () => {
    assert.deepEqual(testPush(TEST_DATA3), TEST_DATA3);
})


/*
Testing the pop stack function
*/
it('Test pop function with TEST_DATA1 and with TEST_DATA_POP1', () => {
    assert.deepEqual(testPop(TEST_DATA1, 3), TEST_DATA_POP1);
})

it('Test pop function with TEST_DATA2 and with TEST_DATA_POP2', () => {
    assert.deepEqual(testPop(TEST_DATA2, 2), TEST_DATA_POP2);
})

it('Test pop function with TEST_DATA3 and with TEST_DATA_POP3', () => {
    assert.deepEqual(testPop(TEST_DATA3, 1), TEST_DATA_POP3);
})


/*
Testing the top stack function
*/
it('Test top function with TEST_DATA1', () => {
    assert.equal(testTop(TEST_DATA1), TEST_DATA1[TEST_DATA1.length - 1]);
})

it('Test top function with TEST_DATA2', () => {
    assert.equal(testTop(TEST_DATA2), TEST_DATA2[TEST_DATA2.length - 1]);
})

it('Test top function with TEST_DATA3', () => {
    assert.equal(testTop(TEST_DATA3), TEST_DATA3[TEST_DATA3.length - 1]);
})

it('Test top function with undefided', () => {
    assert.equal(testTop([]), undefined);
})

File: structures/stack/java/src/test/java/com/bushuevda/stack/stack/StackTest.java
package com.bushuevda.stack.stack;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;

import org.junit.jupiter.api.Test;


public class StackTest {
    /*
    Data for testing all stack functions
    */
    final ArrayList<Integer> TEST_DATA1 = new ArrayList<Integer>(){{
        add(1); add(2); add(3); add(4); add(5); add(6); add(7);
    }};
    final ArrayList<String> TEST_DATA2 = new ArrayList<String>(){{
        add("a"); add("e"); add("b"); add("d"); add("f");}
    };    
    final ArrayList<Integer> TEST_DATA3 = new ArrayList<Integer>(){{
        add(7); add(5); add(1); add(4); add(15); add(6); add(7);
    }};        
     
    /*
    Data for testing stack function pop
    */
    final ArrayList<Integer> TEST_DATA_POP1 = new ArrayList<Integer>(){{  // -3
        add(1); add(2); add(3); add(4);
    }};

    final ArrayList<String> TEST_DATA_POP2 = new ArrayList<String>() {{ // -2
        add("a"); add("e"); add("b");
    }};

    final ArrayList<Integer> TEST_DATA_POP3 = new ArrayList<Integer>(){{ // -1
        add(7); add(5); add(1); add(4); add(15); add(6);
    }};
 
    
    /*
    Testing the swap stack function
    */
    @Test
    public void testSwap() {
        StackTestUtil<Integer> utilTest = new StackTestUtil<Integer>();

        ArrayList<Integer> result = utilTest.testSwap(TEST_DATA1, TEST_DATA3);
        assertArrayEquals(TEST_DATA3.toArray(), result.toArray());

        result = utilTest.testSwap(TEST_DATA3, TEST_DATA1);
        assertArrayEquals(TEST_DATA1.toArray(), result.toArray());
    }


    /*
    Testing the empty stack function
    */
    @Test
    public void testEmpty(){
        StackTestUtil<Integer> utilTestInteger = new StackTestUtil<Integer>();
        StackTestUtil<String> utilTestString = new StackTestUtil<String>();
        assertFalse(utilTestInteger.testEmpty(TEST_DATA1));
        assertFalse(utilTestString.testEmpty(TEST_DATA2));
        assertFalse(utilTestInteger.testEmpty(TEST_DATA3));
        assertTrue(utilTestInteger.testEmpty(new ArrayList<Integer>()));
    }

    /*
    Testing the size stack function
    */
    @Test
    public void testSize(){
        StackTestUtil<Integer> utilTestInteger = new StackTestUtil<Integer>();
        StackTestUtil<String> utilTestString = new StackTestUtil<String>();
        assertEquals(utilTestInteger.testSize(TEST_DATA1), TEST_DATA1.size());
        assertEquals(utilTestString.testSize(TEST_DATA2), TEST_DATA2.size());
        assertEquals(utilTestInteger.testSize(TEST_DATA3), TEST_DATA3.size());
    }

    /*
    Testing the push stack function
    */
    @Test 
    public void testPush(){
        StackTestUtil<Integer> utilTestInteger = new StackTestUtil<Integer>();
        StackTestUtil<String> utilTestString = new StackTestUtil<String>();
        assertArrayEquals(utilTestInteger.testPush(TEST_DATA1).toArray(), TEST_DATA1.toArray());
        assertArrayEquals(utilTestString.testPush(TEST_DATA2).toArray(), TEST_DATA2.toArray());
        assertArrayEquals(utilTestInteger.testPush(TEST_DATA3).toArray(), TEST_DATA3.toArray());
    }

    /*
    Testing the pop stack function
    */
    @Test
    public void testPop(){
        StackTestUtil<Integer> utilTestInteger = new StackTestUtil<Integer>();
        StackTestUtil<String> utilTestString = new StackTestUtil<String>();
        assertArrayEquals(utilTestInteger.testPop(TEST_DATA1, 3).toArray(), TEST_DATA_POP1.toArray());
        assertArrayEquals(utilTestString.testPop(TEST_DATA2, 2).toArray(), TEST_DATA_POP2.toArray());
        assertArrayEquals(utilTestInteger.testPop(TEST_DATA3, 1).toArray(), TEST_DATA_POP3.toArray());
    }

    /*
    Testing the top stack function
    */
    @Test
    public void testTop(){
        StackTestUtil<Integer> utilTestInteger = new StackTestUtil<Integer>();
        StackTestUtil<String> utilTestString = new StackTestUtil<String>();
        assertEquals(utilTestInteger.testTop(TEST_DATA1), TEST_DATA1.get(TEST_DATA1.size() - 1));
        assertEquals(utilTestString.testTop(TEST_DATA2), TEST_DATA2.get(TEST_DATA2.size() - 1));
        assertEquals(utilTestInteger.testTop(TEST_DATA3), TEST_DATA3.get(TEST_DATA3.size() - 1));
    }

}
